plugins/snapd-uefi/fu-snapd-uefi-plugin.c

```diff
@@ -0,0 +1,309 @@
+
+	/* collect response for debugging */
+	(void)curl_easy_setopt(curl, CURLOPT_WRITEDATA, rsp_buf);
+	(void)curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, fu_snapd_uefi_plugin_rsp_cb);
+
+	res = curl_easy_perform(curl);
+	if (res != CURLE_OK) {
+		/* TODO inspect the error */
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "failed to communicate with snapd: %s",
+			    curl_easy_strerror(res));
+		return FALSE;
+	}
+
+	curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &status_code);
+	if (status_code == 404) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "snapd notification endpoint not supported by snapd API");
+		return FALSE;
+	}
+
+	if (status_code != 200) {
+		g_autofree gchar *rsp = NULL;
+		if (rsp_buf->len > 0) {
+			/* make sure the response is printable */
+			rsp = fu_strsafe((const char *)rsp_buf->data, rsp_buf->len + 1);
+		}
+
+		/* TODO check whether the response is even printable? */
+		g_info("snapd request failed with status %ld, response: %s",
+		       (glong)status_code,
+		       rsp != NULL ? rsp : "<none>");
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "snapd request failed with status %ld",
+			    (glong)status_code);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_snapd_uefi_plugin_is_in_snap(void)
+{
+	return getenv("SNAP") != NULL;
+}
+
+static gboolean
+fu_snapd_uefi_plugin_startup(FuPlugin *plugin, FuProgress *progress, GError **error)
+{
+	FuSnapPlugin *self = FU_SNAPD_UEFI_PLUGIN(plugin);
+	FuContext *ctx = fu_plugin_get_context(plugin);
+	const gchar *startup_msg = "{\"action\":\"efi-secureboot-update-startup\"}";
+	g_autoptr(GError) error_local = NULL;
+
+	/* only enable snapd integration if either running inside a snap or we detect that this is a
+	snapd FDE setup. either of these cases makes snapd integration mandatory */
+	if (!fu_snapd_uefi_plugin_is_in_snap()) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "not run as a snap");
+		return FALSE;
+	}
+	if (!fu_context_has_flag(ctx, FU_CONTEXT_FLAG_FDE_SNAPD)) {
+		g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED, "no snap FDE");
+		return FALSE;
+	}
+
+	/* notify snapd of that the DBX manager has started */
+	if (!fu_snapd_uefi_plugin_simple_req(self,
+					     "/v2/system-secureboot",
+					     startup_msg,
+					     strlen(startup_msg),
+					     &error_local)) {
+		/* unless we got specific error indicating lack of relevant APIs, snapd integration
+		 * is considered to be supported, even if snapd itself cannot be reached */
+		self->snapd_integration_supported =
+		    !g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED);
+		g_info("snapd integration non-functional: %s", error_local->message);
+	} else {
+		g_info("snapd integration enabled");
+		self->snapd_integration_supported = TRUE;
+	}
```

Comment:
and this part probes snapd, so far we've established we're in a snap or we have snapd managed FDE:
- snapd has no endpoint - we still want the plugin to be there to inhibit updates on the devices (hence return TRUE at some point)
- snapd returns an error - yes for the plugin (returns TRUE), no for the updates (which will fail when attempted, maybe it would be better served with an explicit if ?)
- snapd returns no error - yes for the plugin, and yes for the updates, if it fails it fails at update time